### DISCUSSION TITLE
Before and After Retry, and Logging

### DISCUSSION THREAD

**Discussion by estahn** [hynek/stamina#29] [Ideas]
I'd like to log retries similar to what is done in the tenacity documentation.

https://tenacity.readthedocs.io/en/latest/#before-and-after-retry-and-logging

**Comment by hynek** [comment]
ENRICO!?!?!? You here!? How long it's been? 20 years? 😅

Anyhow, I think this is subproblem of #21 but an important addition: there should be a plug _before_ and _after_ a retry.

**Reply by estahn** [reply]
HYNEK, indeed it has. I was gobsmacked when I realised it was you.

In regard to the issue, a simpler resolution might be to set the logger to the built-in Python logger instead of `None` 🤔 

https://github.com/hynek/stamina/blob/27f283e64cd000defe79e962d0f91b9c4c5bc443/src/stamina/_instrumentation.py#L13

**Comment by estahn** [comment]
I made it work for my case. In case anyone else needs this.
```python
import logging

import stamina


logger = logging.getLogger("stamina")


def log_retries(
    attempt: int, idle_for: float, exc: Exception, name: str, args: tuple[object, ...], kwargs: dict[str, object]
) -> None:
    logger.warning(
        "stamina.retry_scheduled",
        extra=dict(attempt=attempt, idle_for=idle_for, exc=repr(exc), origin=name),
    )


stamina._instrumentation.INSTRUMENTS.append(log_retries)
```

**Comment by hynek** [comment]
This has been fixed with #35 – the snippet won't work in stamina 23.1 and later.
